hibernate-core/src/main/java/org/hibernate/sql/ast/spi/SqlExpressionResolver.java

```diff
@@ -126,7 +127,12 @@ final class ColumnReferenceKey {
 		public ColumnReferenceKey(String tableQualifier, SelectablePath selectablePath, JdbcMapping jdbcMapping) {
 			this.tableQualifier = tableQualifier;
 			this.selectablePath = selectablePath;
-			this.jdbcMapping = jdbcMapping;
+			if ( jdbcMapping instanceof DiscriminatorTypeImpl discriminatorType ) {
```

Comment:
This code path is pretty hot, so I would rather have you do this check at the use site, wherever that is. Also, other code checks for the fetchable being `DiscriminatorMapping` instead.

Reply:
@beikov https://github.com/hibernate/hibernate-orm/pull/9217/commits/e1841106ceb2c86c4a9e3c7e8cf0a0b4afdb253f